src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs

```diff
@@ -160,46 +254,512 @@ private static bool TryGetCompatibleWindowsRuntimeTypesForVariantType(INamedType
+                    if (genericStartIdx != -1)
+                    {
+                        interfaceStaticsMethod += @interface[genericStartIdx..@interface.Length];
+                    }
+
+                    source.AppendLine($$"""
+                                                new global::System.Runtime.InteropServices.ComWrappers.ComInterfaceEntry
+                                                {
+                                                    IID = global::ABI.{{interfaceStaticsMethod}}.IID,
+                                                    Vtable = global::ABI.{{interfaceStaticsMethod}}.AbiToProjectionVftablePtr
+                                                },
+                                    """);
+                }
+                source.AppendLine($$"""
+                                };
+                                """);
+            }
+            else
+            {
+                source.AppendLine($$"""
+                                        return global::System.Array.Empty<global::System.Runtime.InteropServices.ComWrappers.ComInterfaceEntry>();
+                                """);
+            }
+
+            return source.ToString();
+        }
+
+        internal static void GenerateVtableAttributes(Action<string, string> addSource, ImmutableArray<VtableAttribute> vtableAttributes)
+        {
+            // Using ToImmutableHashSet to avoid duplicate entries from the use of partial classes by the developer
+            // to split out their implementation.  When they do that, we will get multiple entries here for that
+            // and try to generate the same attribute and file with the same data as we use the semantic model
+            // to get all the symbol data rather than the data at an instance of a partial class definition.
+            foreach (var vtableAttribute in vtableAttributes.ToImmutableHashSet())
+            {
+                if (vtableAttribute.Interfaces.Any())
+                {
+                    StringBuilder source = new();
+                    source.AppendLine("using static WinRT.TypeExtensions;\n");
+                    if (!vtableAttribute.IsGlobalNamespace)
+                    {
+                        source.AppendLine($$"""
                         namespace {{vtableAttribute.Namespace}}
                         {
                         """);
-                }
+                    }
 
-                source.Append($$"""
-                    [global::WinRT.WinRTExposedType({{vtableEntries}})]
+                    source.AppendLine($$"""
+                    [global::WinRT.WinRTExposedType(typeof({{vtableAttribute.ClassName}}WinRTTypeDetails))]
                     partial class {{vtableAttribute.ClassName}}
                     {
                     }
+
+                    internal sealed class {{vtableAttribute.ClassName}}WinRTTypeDetails : global::WinRT.IWinRTExposedTypeDetails

```

Comment:
As a follow up, should consider checking whether C# 11 is enabled in the project (via `((CSharpCompilation)compilation).LanguageVersion`) and make all of these types `file` accessible instead of `internal`? For context, the `ComWrappers` COM generators also do the same, as these types are only needed for attributes over types and are never directly referenced anywhere else.

Reply:
I did see the COM generators using file accessible but didn't think to dynamically support it by checking the lang version.  I will look at doing this in an upcoming PR.